!TaskIT

Expressing and managing concurrent computations is indeed a concern of importance to develop applications that scale. A web application may want to use different processes for each of its incoming requests. Or maybe it wants to use a "thread pool" in some cases. In other case, our desktop application may want to send computations to a worker to not block the UI thread. TaskIT is a library that eases the usage of processes in Pharo. It provides abstractions to schedule and/or parallelize of the execution of pieces of code.

In this chapter we will explore the first version of TaskIT. We will at first familiarize the reader with the TaskIT abstractions, guided by examples and code snippets. At the end of the chapter we will discuss some of TaskIT extension points and possible customizations.

!Getting Started

In TaskIT A task is a first class representation of a piece of code to be executed, the main abstraction inside TaskIT. By reifying a task, TaskIT can paralellize it, start it and/or stop it whenever we want. A task is created from a block with a piece of code to be run. The result of the task's execution will be the result of its block. Additionally, we can specify a priority that will be used to decide the order of execution between it and other tasks. If we do not specify a priority explicitly, the default ==userBackgroundPriority== defined in Pharo's ==ProcessorScheduler== is used. We will discuss about the importance of a task's priority in the task runners' section. A task, instance of ==TKTTask==, is created in the following way:

[[[
TKTTask for: [ someObject someMessage ].
TKTTask for: [ someObject someMessage ] at: aPriority.
]]]

Alternatively, blocks are extended to provided a nice syntax sugar to create tasks:

[[[
[ someObject someMessage ] asTask.
[ someObject someMessage ] asTaskAt: aPriority.
]]]

Tasks created in this way are not yet executed. We will discuss the different strategies for executing such tasks in the following sections.

!!Executing a Task

The most basic way to execute a task is to run it as a one shoot task. A one shoot task is a task that is run in a new Pharo process. The Pharo process is discarded after the task is executed. To shoot a task we need to send it the message ==shootIt==.

[[[
[ 2 +  2 ] asTask shootIt.
]]]

or as well with the shortcut,

[[[
[ 2 + 2 ] shootIt.
]]]

Tipically, the Pharo processes do not let us handle it's return value. The process is removed from the system once finished, forbidding us to obtain the result of it's execution. TaskIT provides us with a way to obtain the result of our processes in many different ways by using futures.

!!Obtaining a task's result with futures

As the execution of a task can last some time, the immediate result of the ==shootIt== message is a future object. A future object, instance of ==TKTFuture==, is the promise of an execution's result. A future can hold the value of the finished execution, an error produced by the task execution, or neither if the task has not yet finished.

[[[
aFuture := [ 2 + 2 ] shootIt.
]]]

Once we have a future in our hands, we can retrieve its value both synchronously and asynchronously, and interact with it in many forms.

!!!Synchronous result retrieval

The simplest way to interact with a future is synchronously. That is, when asking for a future's value, it will block the actual process until the value is available. We can do that by sending our future the message ==value==.

[[[
future := [ 2 + 2 ] shootIt.
self assert: future value equals: 4.
]]]

However, it could have happened that the finished in an erroneous state, with an exception. In such case, the exception that was thrown inside the task's execution is forwarded to the sender of ==value==.

[[[
future := [ SomeError signal ] shootIt.
[ future value ] on: SomeError do: [ :error | "We handle the error" ].
]]]

A future can also tell us if the task is already finished or not, by sending it the message ==isValueAvailable==. The ==isValueAvailable== message, opposedly to the ==value== message, will not block the caller's process but return immediately a boolean informing if the task has finished.

[[[
future := [ 2 + 2 ] shootIt.
future isValueAvailable.
]]]

However, waiting synchronously or polling for the task to be finished can be a waste of CPU time sometimes. For those cases when completely synchronous execution does not fit, TaskIT provides an alternative of retrieving a value with a timeout option, using the ==valueTimeoutMilliseconds:== message. When we specify a timeout, we can also provide a block to handle the timeout case using the ==valueTimeoutMilliseconds:ifTimeout:==. If we choose not to provide such a block, the default behavior in case of timeout is to throw a ==TKTTimeoutError== exception.

[[[
future := [ (Delay forMilliseconds: 100) wait ] shootIt.

future
    valueTimeoutMilliseconds: 2
    ifTimeout: [ "if it times out we execute this block"].

future valueTimeoutMilliseconds: 2.
]]]

!!!Aynchronous result retrieval and callbacks

Support for asynchronous execution is provided via callbacks. A callback is a block with zero or one argument that will be executed when the task execution is finished. Futures accept callbacks for success and error of execution. On success, TaskIT will pass to the callback the result of the execution as argument. On the other side, when an error occurs during the task execution, TaskIT will execute the error callback with the exception object. A callback is represented by a ==Block== object and configured for a future object via the ==onSuccess:== and ==onError:== messages. Note that callback arguments are optional, and many callbacks can be configured at the same time for the same future object.

@@comment you should explain because this is strange to have the two onSuccess: following each other

[[[
future := [ 40000000 factorial ] shootIt.
future onSuccess: [ 'Finished factorial and I dont care about the result' logCr ].

future := [ 40000000 factorial ] shootIt.
future onSuccess: [ :result | result logCr ].

future := [ self error ] shootIt.
future onError: [ 'An error ocurred!!' logCr ].

future := [ self error ] shootIt.
future onError: [ :exception | exception asString logCr ].
]]]

Callbacks handle transparently the finalization of a task. On the one hand, the callbacks registered while the task is running (and so, not yet finished) will be saved and executed when the task finishes on the process that executes the task. On the other hand, callbacks that are registered when the task has already finished will run synchronously in the caller process with the already available result.